File: src/backend/src/main/java/br/com/grupopipa/gestaointegrada/atendimento/agendamento/agendamento/AgendamentoService.java
package br.com.grupopipa.gestaointegrada.atendimento.agendamento.agendamento;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import br.com.grupopipa.gestaointegrada.atendimento.agendamento.agendamento.dto.AgendamentoDTO;
import br.com.grupopipa.gestaointegrada.atendimento.agendamento.agendamento.dto.AgendamentoGridDTO;
import br.com.grupopipa.gestaointegrada.atendimento.agendamento.agendamento.dto.SlotDTO;
import br.com.grupopipa.gestaointegrada.core.service.CrudService;

public interface AgendamentoService
        extends CrudService<AgendamentoDTO, AgendamentoGridDTO> {

    List<SlotDTO> listarSlots(UUID agendaId, LocalDate dataInicio, LocalDate dataFim);

    List<AgendamentoDTO> listarConflitosParaPaciente(
            UUID pacienteId, LocalDate dataInicio, LocalDate dataFim);

    AgendamentoDTO cancelar(UUID id);
    AgendamentoDTO realizar(UUID id);
}

File: src/frontend/src/app/components/atendimento/agendamento/agendamento/agendamento.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { plainToInstance } from 'class-transformer';
import { MessageService } from '../../../base/messages/messages.service';
import { BaseService } from '../../../base/base-service';
import { AgendamentoDTO } from './model/agendamento-dto';
import { AgendamentoGridDTO } from './model/agendamento-grid-dto';
import { SlotDTO } from './model/slot-dto';
import { Response } from '../../../base/model/response';

@Injectable()
export class AgendamentoService extends BaseService<AgendamentoDTO, AgendamentoGridDTO> {
  private static readonly DOMINIO = 'agendamento/agendamento';

  constructor() {
    super(inject(HttpClient), inject(MessageService));
  }

  protected override convertToDto(body: unknown): AgendamentoDTO {
    return plainToInstance(AgendamentoDTO, body as object) as AgendamentoDTO;
  }

  protected override convertToGrid(item: AgendamentoGridDTO): AgendamentoGridDTO {
    return plainToInstance(AgendamentoGridDTO, item as object) as AgendamentoGridDTO;
  }

  getDomain(): string {
    return AgendamentoService.DOMINIO;
  }

  listarSlots(agendaId: string, dataInicio: string, dataFim: string): Observable<Response<SlotDTO[]>> {
    return this.httpClient.get<Response<SlotDTO[]>>(
      this.getUrl('/slots'),
      { headers: this.getHeaders(), params: { agendaId, dataInicio, dataFim } }
    );
  }

  conflitoPaciente(
    pacienteId: string, dataInicio: string, dataFim: string
  ): Observable<Response<AgendamentoDTO[]>> {
    return this.httpClient.get<Response<AgendamentoDTO[]>>(
      this.getUrl('/conflito-paciente'),
      { headers: this.getHeaders(), params: { pacienteId, dataInicio, dataFim } }
    );
  }

  listarPorPaciente(
    pessoaId: string, dataInicio: string, dataFim: string
  ): Observable<Response<AgendamentoGridDTO[]>> {
    return this.httpClient.get<Response<AgendamentoGridDTO[]>>(
      this.getUrl('/visao-paciente'),
      { headers: this.getHeaders(), params: { pessoaId, dataInicio, dataFim } }
    );
  }

  vincularAtendimento(agendamentoId: string, atendimentoId: string): Observable<Response<AgendamentoDTO>> {
    return this.httpClient.patch<Response<AgendamentoDTO>>(
      this.getUrl(`/${agendamentoId}/vincular-atendimento/${atendimentoId}`),
      {},
      { headers: this.getHeaders() }
    );
  }

  cancelar(id: string): Observable<Response<AgendamentoDTO>> {
    return this.httpClient.patch<Response<AgendamentoDTO>>(
      this.getUrl('/cancelar/' + id),
      {},
      { headers: this.getHeaders() }
    );
  }

  realizar(id: string): Observable<Response<AgendamentoDTO>> {
    return this.httpClient.patch<Response<AgendamentoDTO>>(
      this.getUrl('/realizar/' + id),
      {},
      { headers: this.getHeaders() }
    );
  }
}
